src/middlewared/middlewared/plugins/rsync.py

```diff
@@ -343,3 +346,165 @@ class Config:
     @accepts(Int('id'))
     async def do_delete(self, id):
         return await self.middleware.call('datastore.delete', self._config.datastore, id)
+
+
+class RsyncTaskService(CRUDService):
+
+    class Config:
+        datastore = 'tasks.rsync'
+        datastore_prefix = 'rsync_'
+
+    async def validate_rsync_task(self, data):
+        verrors = ValidationErrors()
+
+        if ' ' in data.get("user"):
+            raise verrors.add("user", "Usernames cannot have spaces")
+
+        if data.get("extra"):
+                data["extra"] = " ".join(data["extra"])
+        else:
+            data["extra"] = ""
+
+        if data.get("month"):
+                if len(data['month']) == 12:
+                    data['month'] = '*'
+                else:
+                    data['month'] = ",".join(data['month'])
+
+        if data.get("dayweek"):
+                if len(data['dayweek']) == 7:
+                    data['dayweek'] = '*'
+                else:
+                    data['dayweek'] = ",".join(data['dayweek'])
+
+        mode = data.get("mode")
+        rmodule = data.get("remotemodule")
+        if mode == 'module' and not rmodule:
+            verrors.add('remotemodule', 'This field is required')
+
+        rpath = data.get("remotepath")
+        if mode == 'ssh' and not rpath:
+            verrors.add('remotepath', 'This field is required')
+
+        username = data.get("user")
+        if mode == 'ssh':
+            rport = data.get("remoteport")
+            if not rport:
+                verrors.add('remoteport', 'This field is required')
+  
+            try:
+                user = await self.middleware.call('user.query', [('username', '=', username)], {'get': True})
```

Comment:
IIRC users could be from directory services, checking the user table doesn't cut it.

We need to find a way to lookup for DS users as well (could be a temporary shim in notifier.py plugin)